modules/bash-commons/src/aws.sh

```diff
@@ -20,36 +20,65 @@ readonly three_hours_in_s=10800
+function aws_check_metadata_availability {
+  version_to_check=${GRUNTWORK_BASH_COMMONS_IMDS_VERSION:-$default_instance_metadata_version}
+  if [[ "${version_to_check}" == "" ]]; then
+    echo "No IMDS version specified, unable to check metadata availability."
+    return 9
+  fi
+
+  if [[ "${version_to_check}" == "2" ]]; then
+    curl_exit_code=$(sudo curl -s -o /dev/null -X PUT ${imdsv2_token_endpoint} -H "X-aws-ec2-metadata-token-ttl-seconds: 10"; echo $?)
+    if [ ${curl_exit_code} -eq 0 ]; then
       default_instance_metadata_version="2"
-    elif [ ${to_token_or_not_to_token} -eq 7 ]; then
+    elif [ ${curl_exit_code} -eq 7 ]; then
       echo "Check for IMDSv2 failed. IMDS endpoint connection refused."
       default_instance_metadata_version="0"
     else
-      finish_code=$(sudo curl -s -o /dev/null $metadata_endpoint; echo $?)
-      if [ ${finish_code} -eq 0 ]; then
-        default_instance_metadata_version="1"
-      elif [ ${finish_code} -eq 7 ]; then
-        echo "Check for IMDSv1 and v2 failed. IMDS endpoint connection refused."
-        default_instance_metadata_version="0"
-      else
-        echo "IMDS endpoint connection failed for an unknown reason with error code: ${finish_code}"
-        default_instance_metadata_version="0"
-      fi
+      echo "IMDS endpoint connection failed for an unknown reason with error code: ${finish_code}"
+      default_instance_metadata_version="0"
     fi
-fi
+  fi
 
-# shellcheck source=./modules/bash-commons/src/assert.sh
-source "$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)/assert.sh"
-# shellcheck source=./modules/bash-commons/src/log.sh
-source "$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)/log.sh"
+  if [[ "${version_to_check}" == "1" ]]; then
+    curl_exit_code=$(sudo curl -s -o /dev/null $metadata_endpoint; echo $?)
+    if [ ${curl_exit_code} -eq 0 ]; then
+      default_instance_metadata_version="1"
+    elif [ ${curl_exit_code} -eq 7 ]; then
+      echo "Check for IMDSv1 and v2 failed. IMDS endpoint connection refused."
+      default_instance_metadata_version="0"
+    else
+      echo "IMDS endpoint connection failed for an unknown reason with error code: ${finish_code}"
+      default_instance_metadata_version="0"
+    fi
+  fi
+  
+  # returns "2" if IMDSv2 is available, "1" if v2 is not but v1 is, returns "0" if neither are available (i.e. endpoint is disabled or blocked)
+  echo $default_instance_metadata_version
+}
+
+# Check if IMDS Metadata Endpoint is available.  This is required for most of the functions in this script.
+imds_available=$(aws_check_metadata_availability)
+if [[ "${imds_available}" == 9 ]]; then
+  echo "No IMDS Version declared.  This should not be possible because this script sets a default of 2.  Check to see if it was unset somewhere."
+elif [[ "${imds_available}" == 0 ]]; then
+  echo "IMDS Metadata Endpoint is not available.  Script cannot continue without the Endpoint."
```

Comment:
Displays an error to assist with troubleshooting if the IMDS Metadata Endpoint is disabled or not available.  Informs the user this is required for the script to run properly.